{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3ba34025",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: numpy in /home/inferno/.local/lib/python3.6/site-packages (1.19.5)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pandas in /home/inferno/.local/lib/python3.6/site-packages (1.1.5)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/inferno/.local/lib/python3.6/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/inferno/.local/lib/python3.6/site-packages (from pandas) (2021.3)\n",
      "Requirement already satisfied: numpy>=1.15.4 in /home/inferno/.local/lib/python3.6/site-packages (from pandas) (1.19.5)\n",
      "Requirement already satisfied: six>=1.5 in /home/inferno/.local/lib/python3.6/site-packages (from python-dateutil>=2.7.3->pandas) (1.15.0)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: opencv-python in /home/inferno/.local/lib/python3.6/site-packages (4.5.5.62)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /home/inferno/.local/lib/python3.6/site-packages (from opencv-python) (1.19.5)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting tensorflow==2.4\n",
      "  Downloading tensorflow-2.4.0-cp36-cp36m-manylinux2010_x86_64.whl (394.7 MB)\n",
      "     |████████████████████████████████| 394.7 MB 4.5 kB/s             \n",
      "\u001b[?25hRequirement already satisfied: google-pasta~=0.2 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (0.2.0)\n",
      "Requirement already satisfied: keras-preprocessing~=1.1.2 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (1.1.2)\n",
      "Requirement already satisfied: protobuf>=3.9.2 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (3.19.4)\n",
      "Collecting tensorflow-estimator<2.5.0,>=2.4.0rc0\n",
      "  Downloading tensorflow_estimator-2.4.0-py2.py3-none-any.whl (462 kB)\n",
      "     |████████████████████████████████| 462 kB 9.3 MB/s            \n",
      "\u001b[?25hRequirement already satisfied: wheel~=0.35 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (0.37.1)\n",
      "Requirement already satisfied: absl-py~=0.10 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (0.15.0)\n",
      "Requirement already satisfied: termcolor~=1.1.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (1.1.0)\n",
      "Requirement already satisfied: six~=1.15.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (1.15.0)\n",
      "Requirement already satisfied: opt-einsum~=3.3.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (3.3.0)\n",
      "Requirement already satisfied: typing-extensions~=3.7.4 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (3.7.4.3)\n",
      "Requirement already satisfied: wrapt~=1.12.1 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (1.12.1)\n",
      "Requirement already satisfied: flatbuffers~=1.12.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (1.12)\n",
      "Requirement already satisfied: numpy~=1.19.2 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (1.19.5)\n",
      "Collecting gast==0.3.3\n",
      "  Downloading gast-0.3.3-py2.py3-none-any.whl (9.7 kB)\n",
      "Collecting grpcio~=1.32.0\n",
      "  Downloading grpcio-1.32.0-cp36-cp36m-manylinux2014_x86_64.whl (3.8 MB)\n",
      "     |████████████████████████████████| 3.8 MB 12.3 MB/s            \n",
      "\u001b[?25hRequirement already satisfied: tensorboard~=2.4 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (2.6.0)\n",
      "Requirement already satisfied: astunparse~=1.6.3 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow==2.4) (1.6.3)\n",
      "Collecting h5py~=2.10.0\n",
      "  Downloading h5py-2.10.0-cp36-cp36m-manylinux1_x86_64.whl (2.9 MB)\n",
      "     |████████████████████████████████| 2.9 MB 10.1 MB/s            \n",
      "\u001b[?25hRequirement already satisfied: google-auth<2,>=1.6.3 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow==2.4) (1.35.0)\n",
      "Requirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow==2.4) (1.8.1)\n",
      "Requirement already satisfied: tensorboard-data-server<0.7.0,>=0.6.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow==2.4) (0.6.1)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow==2.4) (3.3.6)\n",
      "Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow==2.4) (0.4.6)\n",
      "Requirement already satisfied: werkzeug>=0.11.15 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow==2.4) (2.0.3)\n",
      "Requirement already satisfied: setuptools>=41.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow==2.4) (59.6.0)\n",
      "Requirement already satisfied: requests<3,>=2.21.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow==2.4) (2.27.1)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /home/inferno/.local/lib/python3.6/site-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow==2.4) (4.8)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/inferno/.local/lib/python3.6/site-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow==2.4) (0.2.8)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow==2.4) (4.2.4)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /home/inferno/.local/lib/python3.6/site-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.4->tensorflow==2.4) (1.3.1)\n",
      "Requirement already satisfied: importlib-metadata>=4.4 in /home/inferno/.local/lib/python3.6/site-packages (from markdown>=2.6.8->tensorboard~=2.4->tensorflow==2.4) (4.8.3)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/inferno/.local/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow==2.4) (1.26.8)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/inferno/.local/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow==2.4) (2021.10.8)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow==2.4) (2.0.12)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/inferno/.local/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow==2.4) (3.3)\n",
      "Requirement already satisfied: dataclasses in /home/inferno/.local/lib/python3.6/site-packages (from werkzeug>=0.11.15->tensorboard~=2.4->tensorflow==2.4) (0.8)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/inferno/.local/lib/python3.6/site-packages (from importlib-metadata>=4.4->markdown>=2.6.8->tensorboard~=2.4->tensorflow==2.4) (3.6.0)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /home/inferno/.local/lib/python3.6/site-packages (from pyasn1-modules>=0.2.1->google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow==2.4) (0.4.8)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.4->tensorflow==2.4) (3.2.0)\n",
      "Installing collected packages: grpcio, tensorflow-estimator, h5py, gast, tensorflow\n",
      "  Attempting uninstall: grpcio\n",
      "    Found existing installation: grpcio 1.44.0\n",
      "    Uninstalling grpcio-1.44.0:\n",
      "      Successfully uninstalled grpcio-1.44.0\n",
      "  Attempting uninstall: tensorflow-estimator\n",
      "    Found existing installation: tensorflow-estimator 2.6.0\n",
      "    Uninstalling tensorflow-estimator-2.6.0:\n",
      "      Successfully uninstalled tensorflow-estimator-2.6.0\n",
      "  Attempting uninstall: h5py\n",
      "    Found existing installation: h5py 3.1.0\n",
      "    Uninstalling h5py-3.1.0:\n",
      "      Successfully uninstalled h5py-3.1.0\n",
      "  Attempting uninstall: gast\n",
      "    Found existing installation: gast 0.4.0\n",
      "    Uninstalling gast-0.4.0:\n",
      "      Successfully uninstalled gast-0.4.0\n",
      "  Attempting uninstall: tensorflow\n",
      "    Found existing installation: tensorflow 2.6.2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Uninstalling tensorflow-2.6.2:\n",
      "      Successfully uninstalled tensorflow-2.6.2\n",
      "Successfully installed gast-0.3.3 grpcio-1.32.0 h5py-2.10.0 tensorflow-2.4.0 tensorflow-estimator-2.4.0\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: fastapi in /home/inferno/.local/lib/python3.6/site-packages (0.74.1)\n",
      "Requirement already satisfied: pydantic!=1.7,!=1.7.1,!=1.7.2,!=1.7.3,!=1.8,!=1.8.1,<2.0.0,>=1.6.2 in /home/inferno/.local/lib/python3.6/site-packages (from fastapi) (1.9.0)\n",
      "Requirement already satisfied: starlette==0.17.1 in /home/inferno/.local/lib/python3.6/site-packages (from fastapi) (0.17.1)\n",
      "Requirement already satisfied: contextlib2>=21.6.0 in /home/inferno/.local/lib/python3.6/site-packages (from starlette==0.17.1->fastapi) (21.6.0)\n",
      "Requirement already satisfied: typing-extensions in /home/inferno/.local/lib/python3.6/site-packages (from starlette==0.17.1->fastapi) (3.7.4.3)\n",
      "Requirement already satisfied: anyio<4,>=3.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from starlette==0.17.1->fastapi) (3.5.0)\n",
      "Requirement already satisfied: dataclasses>=0.6 in /home/inferno/.local/lib/python3.6/site-packages (from pydantic!=1.7,!=1.7.1,!=1.7.2,!=1.7.3,!=1.8,!=1.8.1,<2.0.0,>=1.6.2->fastapi) (0.8)\n",
      "Requirement already satisfied: idna>=2.8 in /home/inferno/.local/lib/python3.6/site-packages (from anyio<4,>=3.0.0->starlette==0.17.1->fastapi) (3.3)\n",
      "Requirement already satisfied: contextvars in /home/inferno/.local/lib/python3.6/site-packages (from anyio<4,>=3.0.0->starlette==0.17.1->fastapi) (2.4)\n",
      "Requirement already satisfied: sniffio>=1.1 in /home/inferno/.local/lib/python3.6/site-packages (from anyio<4,>=3.0.0->starlette==0.17.1->fastapi) (1.2.0)\n",
      "Requirement already satisfied: immutables>=0.9 in /home/inferno/.local/lib/python3.6/site-packages (from contextvars->anyio<4,>=3.0.0->starlette==0.17.1->fastapi) (0.16)\n"
     ]
    }
   ],
   "source": [
    "!pip install numpy\n",
    "!pip install pandas\n",
    "!pip install opencv-python\n",
    "!pip install tensorflow==2.4\n",
    "!pip install fastapi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2926a9c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pillow in /home/inferno/.local/lib/python3.6/site-packages (8.4.0)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pefile in /home/inferno/.local/lib/python3.6/site-packages (2021.9.3)\n",
      "Requirement already satisfied: future in /home/inferno/.local/lib/python3.6/site-packages (from pefile) (0.18.2)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting scipy==1.3\n",
      "  Downloading scipy-1.3.0-cp36-cp36m-manylinux1_x86_64.whl (25.2 MB)\n",
      "     |████████████████████████████████| 25.2 MB 5.9 MB/s            \n",
      "\u001b[?25hRequirement already satisfied: numpy>=1.13.3 in /home/inferno/.local/lib/python3.6/site-packages (from scipy==1.3) (1.19.5)\n",
      "Installing collected packages: scipy\n",
      "  Attempting uninstall: scipy\n",
      "    Found existing installation: scipy 1.5.4\n",
      "    Uninstalling scipy-1.5.4:\n",
      "      Successfully uninstalled scipy-1.5.4\n",
      "Successfully installed scipy-1.3.0\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting scikit-learn==0.19.1\n",
      "  Downloading scikit_learn-0.19.1-cp36-cp36m-manylinux1_x86_64.whl (12.4 MB)\n",
      "     |████████████████████████████████| 12.4 MB 5.8 MB/s            \n",
      "\u001b[?25hInstalling collected packages: scikit-learn\n",
      "  Attempting uninstall: scikit-learn\n",
      "    Found existing installation: scikit-learn 0.19.0\n",
      "    Uninstalling scikit-learn-0.19.0:\n",
      "      Successfully uninstalled scikit-learn-0.19.0\n",
      "Successfully installed scikit-learn-0.19.1\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: uvicorn in /home/inferno/.local/lib/python3.6/site-packages (0.13.1)\n",
      "Requirement already satisfied: typing-extensions in /home/inferno/.local/lib/python3.6/site-packages (from uvicorn) (3.7.4.3)\n",
      "Requirement already satisfied: h11>=0.8 in /home/inferno/.local/lib/python3.6/site-packages (from uvicorn) (0.13.0)\n",
      "Requirement already satisfied: click==7.* in /home/inferno/.local/lib/python3.6/site-packages (from uvicorn) (7.1.2)\n",
      "Requirement already satisfied: dataclasses in /home/inferno/.local/lib/python3.6/site-packages (from h11>=0.8->uvicorn) (0.8)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: python-multipart in /home/inferno/.local/lib/python3.6/site-packages (0.0.5)\n",
      "Requirement already satisfied: six>=1.4.0 in /home/inferno/.local/lib/python3.6/site-packages (from python-multipart) (1.15.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install pillow\n",
    "!pip install pefile\n",
    "!pip install scipy==1.3\n",
    "!pip install scikit-learn==0.19.1\n",
    "!pip install uvicorn\n",
    "!pip install python-multipart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0bbf8d16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: aiofiles in /home/inferno/.local/lib/python3.6/site-packages (0.8.0)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: jinja2 in /home/inferno/.local/lib/python3.6/site-packages (3.0.3)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/inferno/.local/lib/python3.6/site-packages (from jinja2) (2.0.1)\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting keras==2.4\n",
      "  Downloading Keras-2.4.0-py2.py3-none-any.whl (170 kB)\n",
      "     |████████████████████████████████| 170 kB 7.2 MB/s            \n",
      "\u001b[?25hRequirement already satisfied: tensorflow>=2.2.0 in /home/inferno/.local/lib/python3.6/site-packages (from keras==2.4) (2.4.0)\n",
      "Requirement already satisfied: numpy>=1.9.1 in /home/inferno/.local/lib/python3.6/site-packages (from keras==2.4) (1.19.5)\n",
      "Requirement already satisfied: scipy>=0.14 in /home/inferno/.local/lib/python3.6/site-packages (from keras==2.4) (1.3.0)\n",
      "Requirement already satisfied: h5py in /home/inferno/.local/lib/python3.6/site-packages (from keras==2.4) (2.10.0)\n",
      "Requirement already satisfied: pyyaml in /home/inferno/.local/lib/python3.6/site-packages (from keras==2.4) (6.0)\n",
      "Requirement already satisfied: keras-preprocessing~=1.1.2 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (1.1.2)\n",
      "Requirement already satisfied: wheel~=0.35 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (0.37.1)\n",
      "Requirement already satisfied: gast==0.3.3 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (0.3.3)\n",
      "Requirement already satisfied: protobuf>=3.9.2 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (3.19.4)\n",
      "Requirement already satisfied: termcolor~=1.1.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (1.1.0)\n",
      "Requirement already satisfied: wrapt~=1.12.1 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (1.12.1)\n",
      "Requirement already satisfied: typing-extensions~=3.7.4 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (3.7.4.3)\n",
      "Requirement already satisfied: absl-py~=0.10 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (0.15.0)\n",
      "Requirement already satisfied: grpcio~=1.32.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (1.32.0)\n",
      "Requirement already satisfied: flatbuffers~=1.12.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (1.12)\n",
      "Requirement already satisfied: tensorflow-estimator<2.5.0,>=2.4.0rc0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (2.4.0)\n",
      "Requirement already satisfied: google-pasta~=0.2 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (0.2.0)\n",
      "Requirement already satisfied: six~=1.15.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (1.15.0)\n",
      "Requirement already satisfied: tensorboard~=2.4 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (2.6.0)\n",
      "Requirement already satisfied: astunparse~=1.6.3 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (1.6.3)\n",
      "Requirement already satisfied: opt-einsum~=3.3.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorflow>=2.2.0->keras==2.4) (3.3.0)\n",
      "Requirement already satisfied: requests<3,>=2.21.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (2.27.1)\n",
      "Requirement already satisfied: tensorboard-data-server<0.7.0,>=0.6.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (0.6.1)\n",
      "Requirement already satisfied: google-auth<2,>=1.6.3 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (1.35.0)\n",
      "Requirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (1.8.1)\n",
      "Requirement already satisfied: setuptools>=41.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (59.6.0)\n",
      "Requirement already satisfied: werkzeug>=0.11.15 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (2.0.3)\n",
      "Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (0.4.6)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /home/inferno/.local/lib/python3.6/site-packages (from tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (3.3.6)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/inferno/.local/lib/python3.6/site-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (0.2.8)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /home/inferno/.local/lib/python3.6/site-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (4.8)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (4.2.4)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /home/inferno/.local/lib/python3.6/site-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (1.3.1)\n",
      "Requirement already satisfied: importlib-metadata>=4.4 in /home/inferno/.local/lib/python3.6/site-packages (from markdown>=2.6.8->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (4.8.3)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (2.0.12)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/inferno/.local/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (2021.10.8)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/inferno/.local/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (3.3)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/inferno/.local/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (1.26.8)\n",
      "Requirement already satisfied: dataclasses in /home/inferno/.local/lib/python3.6/site-packages (from werkzeug>=0.11.15->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (0.8)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/inferno/.local/lib/python3.6/site-packages (from importlib-metadata>=4.4->markdown>=2.6.8->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (3.6.0)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /home/inferno/.local/lib/python3.6/site-packages (from pyasn1-modules>=0.2.1->google-auth<2,>=1.6.3->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (0.4.8)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /home/inferno/.local/lib/python3.6/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.4->tensorflow>=2.2.0->keras==2.4) (3.2.0)\n",
      "Installing collected packages: keras\n",
      "  Attempting uninstall: keras\n",
      "    Found existing installation: keras 2.6.0\n",
      "    Uninstalling keras-2.6.0:\n",
      "      Successfully uninstalled keras-2.6.0\n",
      "Successfully installed keras-2.4.0\n"
     ]
    }
   ],
   "source": [
    "!pip install aiofiles\n",
    "!pip install jinja2\n",
    "!pip install keras==2.4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d1149740",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/inferno/.local/lib/python3.6/site-packages/pandas/compat/__init__.py:120: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.\n",
      "  warnings.warn(msg)\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import cv2\n",
    "import tensorflow as tf \n",
    "import itertools\n",
    "import pefile\n",
    "import os,math\n",
    "from queue import Queue\n",
    "from threading import Thread\n",
    "import argparse\n",
    "from PIL import Image\n",
    "import uvicorn\n",
    "from keras.preprocessing.image import img_to_array\n",
    "from keras.applications import imagenet_utils\n",
    "import tempfile\n",
    "from fastapi import FastAPI, File, UploadFile\n",
    "from starlette.responses import RedirectResponse\n",
    "from sklearn.externals import joblib\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.pipeline import Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1638b956",
   "metadata": {},
   "outputs": [],
   "source": [
    "#install the pefile module using !pip install pefile\n",
    "\n",
    "def createDataframeFromPEdump(pe):\n",
    "\n",
    "\tdosHeaders = ['e_magic', 'e_cblp', 'e_cp', 'e_crlc', 'e_cparhdr',\n",
    "\t       'e_minalloc', 'e_maxalloc', 'e_ss', 'e_sp', 'e_csum', 'e_ip', 'e_cs',\n",
    "\t       'e_lfarlc', 'e_ovno', 'e_oemid', 'e_oeminfo', 'e_lfanew']\n",
    "\tfileHeaders=['Machine',\n",
    "\t       'NumberOfSections', 'TimeDateStamp', 'PointerToSymbolTable',\n",
    "\t       'NumberOfSymbols', 'SizeOfOptionalHeader', 'Characteristics']\n",
    "\toptionalHeaders=['Magic',\n",
    "\t       'MajorLinkerVersion', 'MinorLinkerVersion', 'SizeOfCode',\n",
    "\t       'SizeOfInitializedData', 'SizeOfUninitializedData','AddressOfEntryPoint', 'BaseOfCode',\n",
    "\t        'ImageBase', 'SectionAlignment','FileAlignment', 'MajorOperatingSystemVersion',\n",
    "\t       'MinorOperatingSystemVersion', 'MajorImageVersion', 'MinorImageVersion',\n",
    "\t       'MajorSubsystemVersion', 'MinorSubsystemVersion', 'SizeOfHeaders',\n",
    "\t       'CheckSum', 'SizeOfImage', 'Subsystem', 'DllCharacteristics',\n",
    "\t       'SizeOfStackReserve', 'SizeOfStackCommit', 'SizeOfHeapReserve',\n",
    "\t       'SizeOfHeapCommit', 'LoaderFlags', 'NumberOfRvaAndSizes']\n",
    "\timageDirectory = ['ImageDirectoryEntryExport', 'ImageDirectoryEntryImport',\n",
    "\t       'ImageDirectoryEntryResource', 'ImageDirectoryEntryException',\n",
    "\t       'ImageDirectoryEntrySecurity']\n",
    "\n",
    "\tdheaders = {}\n",
    "\tfheaders = {}\n",
    "\toheaders={}\n",
    "\timd1 = {}\n",
    "\n",
    "\tfor x in dosHeaders:\n",
    "\t\tdheaders[x] = getattr(pe.DOS_HEADER,x)\n",
    "\tdf = pd.DataFrame(dheaders,index=[0])\n",
    "\n",
    "\tfor i in fileHeaders:\n",
    "\t\tfheaders[i] = getattr(pe.FILE_HEADER,i)\n",
    "\tdf = pd.concat([df,(pd.DataFrame(fheaders,index=[0]))],axis=1)\n",
    "\n",
    "\tfor y in optionalHeaders:\n",
    "\t\toheaders[y] = getattr(pe.OPTIONAL_HEADER,y)\n",
    "\tdf = pd.concat([df,(pd.DataFrame(oheaders,index=[0]))],axis=1)\n",
    "\n",
    "\tfor q in pe.OPTIONAL_HEADER.DATA_DIRECTORY:\n",
    "\t    imd1[q.name] = q.VirtualAddress\n",
    "\timd1 = dict(itertools.islice(imd1.items(),5))\n",
    "\tdf = pd.concat([df,(pd.DataFrame(imd1,index=[0]))],axis=1)\n",
    "\n",
    "\treturn df\n",
    "\n",
    "def getPredictions(df):\n",
    "\tload_scaler = joblib.load(open(r'../components/SScaler.pkl','rb'))\n",
    "\tload_skpca = joblib.load(open(r'../components/PCA.pkl','rb'))\n",
    "\tload_model = joblib.load(open(r'../components/model.pkl','rb'))\n",
    "\tpipe = Pipeline([('scale', load_scaler),('pca', load_skpca),('clf', load_model)])\n",
    "\tdf = np.array(df)\n",
    "\tdf = df.reshape(1,-1)\n",
    "\tresults = pipe.predict_proba(df)\n",
    "\t\n",
    "\treturn results[0]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dbe74e35",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PIL import Image\n",
    "def getBinaryData(filename):\n",
    "\t\"\"\"\n",
    "\tExtract byte values from binary executable file and store them into list\n",
    "\t:param filename: executable file name\n",
    "\t:return: byte value list\n",
    "\t\"\"\"\n",
    "\tbinary_values = []\n",
    "\n",
    "\twith open(filename, 'rb') as fileobject:\n",
    "\n",
    "\t\t# read file byte by byte\n",
    "\t\tdata = fileobject.read(1)\n",
    "\n",
    "\t\twhile data != b'':\n",
    "\t\t\tbinary_values.append(ord(data))\n",
    "\t\t\tdata = fileobject.read(1)\n",
    "\treturn binary_values\n",
    "\n",
    "def createGreyScaleImage(filename, width=None):\n",
    "    global img_path \n",
    "    greyscale_data  = getBinaryData(filename)\n",
    "    size            = get_size(len(greyscale_data), width)\n",
    "    # try:\n",
    "    image = Image.new('L',size)\n",
    "    image.putdata(greyscale_data)\n",
    "      # setup output filename\n",
    "    dirname     = os.path.dirname(filename)\n",
    "    name, _     = os.path.splitext(filename)\n",
    "    name        = os.path.basename(name)\n",
    "    imagename   = dirname + os.sep + name + '_'+ '.png'\n",
    "    os.makedirs(os.path.dirname(imagename), exist_ok=True)\n",
    "    img_path = imagename\n",
    "    image.save(imagename)\n",
    "    print('The file', imagename, 'saved.')\n",
    "    # except Exception as err:\n",
    "    # \t  print(err)\n",
    "  \n",
    "\n",
    "\n",
    "def get_size(data_length, width=None):\n",
    "\t# source Malware images: visualization and automatic classification by L. Nataraj\n",
    "\t# url : http://dl.acm.org/citation.cfm?id=2016908\n",
    "\n",
    "\tif width is None: # with don't specified any with value\n",
    "\n",
    "\t\tsize = data_length\n",
    "\n",
    "\t\tif (size < 10240):\n",
    "\t\t\twidth = 32\n",
    "\t\telif (10240 <= size <= 10240 * 3):\n",
    "\t\t\twidth = 64\n",
    "\t\telif (10240 * 3 <= size <= 10240 * 6):\n",
    "\t\t\twidth = 128\n",
    "\t\telif (10240 * 6 <= size <= 10240 * 10):\n",
    "\t\t\twidth = 256\n",
    "\t\telif (10240 * 10 <= size <= 10240 * 20):\n",
    "\t\t\twidth = 384\n",
    "\t\telif (10240 * 20 <= size <= 10240 * 50):\n",
    "\t\t\twidth = 512\n",
    "\t\telif (10240 * 50 <= size <= 10240 * 100):\n",
    "\t\t\twidth = 768\n",
    "\t\telse:\n",
    "\t\t\twidth = 1024\n",
    "\n",
    "\t\theight = int(size / width) + 1\n",
    "\n",
    "\telse:\n",
    "\t\twidth  = int(math.sqrt(data_length)) + 1\n",
    "\t\theight = width\n",
    "\n",
    "\treturn (width, height)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5d6c4faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dict = {'Adialer.C': 0,'Agent.FYI': 1,'Allaple.A': 2,'Allaple.L': 3,'Alueron.gen!J': 4,'Autorun.K': 5,\n",
    " 'C2LOP.P': 6,'C2LOP.gen!g': 7,'Dialplatform.B': 8,'Dontovo.A': 9, 'Fakerean': 10, 'Instantaccess': 11, \n",
    " 'Lolyda.AA1': 12, 'Lolyda.AA2': 13, 'Lolyda.AA3': 14, 'Lolyda.AT': 15, 'Malex.gen!J': 16,'Obfuscator.AD': 17,\n",
    " 'Rbot!gen': 18,'Skintrim.N': 19, 'Swizzor.gen!E': 20,'Swizzor.gen!I': 21, 'VB.AT': 22,'Wintrim.BX': 23,'Yuner.A': 24}\n",
    "\n",
    "\n",
    "\n",
    "# def preprocessImage(path):\n",
    "#     image = cv2.imread(path)\n",
    "#     image = cv2.resize(image,(64,64))\n",
    "#     image = img_to_array(image)\n",
    "#     image = np.expand_dims(image,axis=0)\n",
    "#     return image\n",
    "\n",
    "# def cnn_predict(image):\n",
    "#     global model\n",
    "#     if model is None:\n",
    "#         model = tf.keras.models.load_model('../components/malware_model.h5')\n",
    "#         print(\"Model Loaded\")\n",
    "# #         results = (model.predict(image) > 0.5).astype(\"int32\")\n",
    "# #         results = model.predict_classes(image)\n",
    "#         predict_x=model.predict(image) \n",
    "#         results=np.argmax(predict_x,axis=-1)\n",
    "\n",
    "model= None\n",
    "\n",
    "def load_model():\n",
    "  model =  tf.keras.models.load_model(r'../components/malware_model.h5')\n",
    "  print(\"Model loaded...\")\n",
    "  return model\n",
    "\n",
    "def cnn_predict(imagepath):\n",
    "    global model\n",
    "    if model is None:\n",
    "        model = load_model()\n",
    "    image = cv2.imread(imagepath)\n",
    "    image = cv2.resize(image, (64, 64))\n",
    "    image = img_to_array(image)\n",
    "    image = np.expand_dims(image, axis=0)\n",
    "#     res = model.predict_classes(image)\n",
    "    ax = model.predict(image)\n",
    "    res = np.argmax(ax , axis = -1)\n",
    "    \n",
    "    return get_key(res)\n",
    "\n",
    "def get_key(val):\n",
    "    for key, value in output_dict.items():\n",
    "         if val == value:\n",
    "             return key\n",
    " \n",
    "    return \"key doesn't exist\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5a66e399",
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "This event loop is already running",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-f8436ebbf22a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     31\u001b[0m     \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m \u001b[0muvicorn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mapp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mport\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m8000\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/uvicorn/main.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(app, **kwargs)\u001b[0m\n\u001b[1;32m    384\u001b[0m         \u001b[0msupervisor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    385\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 386\u001b[0;31m         \u001b[0mserver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    387\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    388\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/uvicorn/server.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, sockets)\u001b[0m\n\u001b[1;32m     46\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetup_event_loop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m         \u001b[0mloop\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0masyncio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_event_loop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 48\u001b[0;31m         \u001b[0mloop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_until_complete\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msockets\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msockets\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     49\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     50\u001b[0m     \u001b[0;34masync\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mserve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msockets\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/asyncio/base_events.py\u001b[0m in \u001b[0;36mrun_until_complete\u001b[0;34m(self, future)\u001b[0m\n\u001b[1;32m    473\u001b[0m         \u001b[0mfuture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_done_callback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_run_until_complete_cb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    474\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 475\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_forever\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    476\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    477\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mnew_task\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mfuture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdone\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mfuture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcancelled\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/asyncio/base_events.py\u001b[0m in \u001b[0;36mrun_forever\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    427\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_check_closed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    428\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_running\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 429\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'This event loop is already running'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    430\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mevents\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_running_loop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    431\u001b[0m             raise RuntimeError(\n",
      "\u001b[0;31mRuntimeError\u001b[0m: This event loop is already running"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:     Started server process [3211]\n",
      "INFO:     Waiting for application startup.\n",
      "INFO:     Application startup complete.\n",
      "INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:     127.0.0.1:49098 - \"GET /docs HTTP/1.1\" 200 OK\n",
      "INFO:     127.0.0.1:49098 - \"GET /openapi.json HTTP/1.1\" 200 OK\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/inferno/.local/lib/python3.6/site-packages/sklearn/utils/validation.py:475: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.\n",
      "  warnings.warn(msg, DataConversionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "it is malware\n",
      "The file /tmp/parser_k75ttgp3_.png saved.\n",
      "Model loaded...\n",
      "INFO:     127.0.0.1:49102 - \"POST /predict HTTP/1.1\" 200 OK\n"
     ]
    }
   ],
   "source": [
    "app_desc = \"\"\"<h2> Try uploading a Portable Executable(PE) file\"\"\"\n",
    "app = FastAPI(description = app_desc)\n",
    "\n",
    "@app.get(\"/\",include_in_schema=False)\n",
    "async def index():\n",
    "\treturn RedirectResponse(url=\"/docs\")\n",
    "\n",
    "@app.post(\"/predict\")\n",
    "def parse(file: UploadFile = File(...)):\n",
    "    extension = os.path.splitext(file.filename)[1]\n",
    "    _, path = tempfile.mkstemp(prefix='parser_', suffix=extension)\n",
    "\n",
    "    with open(path, 'ab') as f:\n",
    "        for chunk in iter(lambda: file.file.read(10000), b''):\n",
    "            f.write(chunk)\n",
    "\n",
    "    # extract content\n",
    "    content = pefile.PE(path,fast_load=True)\n",
    "    img = None\n",
    "    dataframe = createDataframeFromPEdump(content)\n",
    "    binary_preds = getPredictions(dataframe)\n",
    "    if binary_preds[1]*100 > 60.0:\n",
    "      print(\"it is malware\")\n",
    "      createGreyScaleImage(path)\n",
    "      img = cnn_predict(img_path)\n",
    "      return {'response':'OK','path':path,'predictions':binary_preds[1],'malware':img}\n",
    "    else:\n",
    "      return{'Response': 'Your file is safe from malware.','Malicious percentage':binary_preds[1]*100}\n",
    "    # remove temp file\n",
    "    os.close(_)\n",
    "    os.remove(path)\n",
    "\n",
    "uvicorn.run(app,port=8000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5ebc3a55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                 Version\n",
      "----------------------- -------------------\n",
      "absl-py                 0.15.0\n",
      "aiofiles                0.8.0\n",
      "anyio                   3.5.0\n",
      "argon2-cffi             21.3.0\n",
      "argon2-cffi-bindings    21.2.0\n",
      "asgiref                 3.4.1\n",
      "astunparse              1.6.3\n",
      "async-generator         1.10\n",
      "attrs                   21.4.0\n",
      "backcall                0.2.0\n",
      "bleach                  4.1.0\n",
      "cached-property         1.5.2\n",
      "cachetools              4.2.4\n",
      "certifi                 2021.10.8\n",
      "cffi                    1.15.0\n",
      "charset-normalizer      2.0.12\n",
      "clang                   5.0\n",
      "click                   8.0.4\n",
      "contextlib2             21.6.0\n",
      "contextvars             2.4\n",
      "dataclasses             0.8\n",
      "decorator               5.1.1\n",
      "defusedxml              0.7.1\n",
      "entrypoints             0.4\n",
      "fastapi                 0.74.1\n",
      "flatbuffers             1.12\n",
      "future                  0.18.2\n",
      "gast                    0.4.0\n",
      "google-auth             1.35.0\n",
      "google-auth-oauthlib    0.4.6\n",
      "google-pasta            0.2.0\n",
      "grpcio                  1.34.1\n",
      "h11                     0.13.0\n",
      "h5py                    3.1.0\n",
      "idna                    3.3\n",
      "immutables              0.16\n",
      "importlib-metadata      4.8.3\n",
      "ipykernel               5.5.6\n",
      "ipython                 7.16.3\n",
      "ipython-genutils        0.2.0\n",
      "ipywidgets              7.6.5\n",
      "jedi                    0.17.2\n",
      "Jinja2                  3.0.3\n",
      "jsonschema              3.2.0\n",
      "jupyter                 1.0.0\n",
      "jupyter-client          7.1.2\n",
      "jupyter-console         6.4.0\n",
      "jupyter-core            4.9.2\n",
      "jupyterlab-pygments     0.1.2\n",
      "jupyterlab-widgets      1.0.2\n",
      "keras                   2.6.0\n",
      "keras-nightly           2.5.0.dev2021032900\n",
      "Keras-Preprocessing     1.1.2\n",
      "Markdown                3.3.6\n",
      "MarkupSafe              2.0.1\n",
      "mistune                 0.8.4\n",
      "nbclient                0.5.9\n",
      "nbconvert               6.0.7\n",
      "nbformat                5.1.3\n",
      "nest-asyncio            1.5.4\n",
      "notebook                6.4.8\n",
      "numpy                   1.19.5\n",
      "oauthlib                3.2.0\n",
      "opencv-python           4.5.5.62\n",
      "opt-einsum              3.3.0\n",
      "packaging               21.3\n",
      "pandas                  1.1.5\n",
      "pandocfilters           1.5.0\n",
      "parso                   0.7.1\n",
      "pefile                  2021.9.3\n",
      "pexpect                 4.8.0\n",
      "pickleshare             0.7.5\n",
      "Pillow                  8.4.0\n",
      "pip                     21.3.1\n",
      "prometheus-client       0.13.1\n",
      "prompt-toolkit          3.0.28\n",
      "protobuf                3.19.4\n",
      "ptyprocess              0.7.0\n",
      "pyasn1                  0.4.8\n",
      "pyasn1-modules          0.2.8\n",
      "pycparser               2.21\n",
      "pydantic                1.9.0\n",
      "Pygments                2.11.2\n",
      "pyparsing               3.0.7\n",
      "pyrsistent              0.18.0\n",
      "python-dateutil         2.8.2\n",
      "python-multipart        0.0.5\n",
      "pytz                    2021.3\n",
      "pyzmq                   22.3.0\n",
      "qtconsole               5.2.2\n",
      "QtPy                    2.0.1\n",
      "requests                2.27.1\n",
      "requests-oauthlib       1.3.1\n",
      "rsa                     4.8\n",
      "scikit-learn            0.19.1\n",
      "scipy                   1.3.0\n",
      "Send2Trash              1.8.0\n",
      "setuptools              59.6.0\n",
      "six                     1.15.0\n",
      "sniffio                 1.2.0\n",
      "starlette               0.17.1\n",
      "tensorboard             2.6.0\n",
      "tensorboard-data-server 0.6.1\n",
      "tensorboard-plugin-wit  1.8.1\n",
      "tensorflow              2.5.0\n",
      "tensorflow-estimator    2.5.0\n",
      "termcolor               1.1.0\n",
      "terminado               0.12.1\n",
      "testpath                0.6.0\n",
      "tornado                 6.1\n",
      "traitlets               4.3.3\n",
      "typing-extensions       3.7.4.3\n",
      "urllib3                 1.26.8\n",
      "uvicorn                 0.16.0\n",
      "wcwidth                 0.2.5\n",
      "webencodings            0.5.1\n",
      "Werkzeug                2.0.3\n",
      "wheel                   0.37.1\n",
      "widgetsnbextension      3.5.2\n",
      "wrapt                   1.12.1\n",
      "zipp                    3.6.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "898c9cbf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
